### PR TITLE
[SPARK-15792][SQL] Allows operator to change the verbosity in explain output

### PR DESCRIPTION
**Pull request by clockfly**
## What changes were proposed in this pull request?

This PR allows customization of verbosity in explain output. After change, `dataframe.explain()` and `dataframe.explain(true)` has different verbosity output for physical plan. 

Currently, this PR only enables verbosity string for operator `HashAggregateExec` and `SortAggregateExec`. We will gradually enable verbosity string for more operators in future.

**Less verbose mode:** dataframe.explain(extended = false)

`output=[count(a)#85L]` is **NOT** displayed for HashAggregate.

```
scala> Seq((1,2,3)).toDF("a", "b", "c").createTempView("df2")
scala> spark.sql("select count(a) from df2").explain()
== Physical Plan ==
*HashAggregate(key=[], functions=[count(1)])
+- Exchange SinglePartition
   +- *HashAggregate(key=[], functions=[partial_count(1)])
      +- LocalTableScan
```

**Verbose mode:** dataframe.explain(extended = true)

`output=[count(a)#85L]` is displayed for HashAggregate.

```
scala> spark.sql("select count(a) from df2").explain(true)  // "output=[count(a)#85L]" is added
...
== Physical Plan ==
*HashAggregate(key=[], functions=[count(1)], output=[count(a)#85L])
+- Exchange SinglePartition
   +- *HashAggregate(key=[], functions=[partial_count(1)], output=[count#87L])
      +- LocalTableScan
```
## How was this patch tested?

Manual test.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala [hunk 1]
```diff
@@ -201,24 +201,26 @@ class QueryExecution(val sparkSession: SparkSession, val logical: LogicalPlan) {
...
+      stringOrError(output),
+      stringOrError(analyzed.treeString(verbose = true))
+    ).filter(_.nonEmpty).mkString("\n")
 
     s"""== Parsed Logical Plan ==
-       |${stringOrError(logical)}
+       |${stringOrError(logical.treeString(verbose = true))}
        |== Analyzed Logical Plan ==
-       |$analyzedPlan
+       |${analyzedPlan}
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala]: Please revert this change. The braces are unnecessary.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 2]
```diff
@@ -190,6 +190,9 @@ abstract class Expression extends TreeNode[Expression] {
     case single => single :: Nil
   }
 
+  // Marks this as final, verboseString of Expression is NEVER called.
+  final override def verboseString: String = simpleString
+
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: > Marks this as final, `Expression.verboseString` should never be called, and thus shouldn't be overriden by concrete classes.

**Comment by liancheng** [comment]
The PR description says the verbosity mark affects output of "physical plan", but actually only aggregate operators are affected. Please correct the PR description and add that explain output with different verbosity for all physical plans will be added in follow-ups.

**Comment by clockfly** [comment]
@liancheng  updated.

**Comment by liancheng** [comment]
LGTM, merging to master and branch-2.0. Thanks!
